### ISSUE TITLE
Your Memcached instance may be publicly accessible

### ISSUE THREAD

**Issue by fshmcallister** [WoTTsecurity/wott-io#211]
We detected that a Memcached instance on {devices} may be accessible remotely. Consider either blocking port 11211 through the WoTT firewall management tool, or re-configure Memcached to only listen on localhost.

part of #198

**Comment by fshmcallister** [comment]
**TITLE**
A Memcached instance may publicly accessible 
> reconfigure Memcached 


**TL;DR**
A Memcached instance on [DEVICE(S)] may be accessible remotely. This is likely due to use of a default configuration. Disable `UDP` connections and traffic by blocking port 11211. Alternatively, reconfigure Memcached to only listen on localhost. 

**FAQ**
Memcached is a popular caching service used to help speed up websites. By default, Memcached enables TCP and UDP protocols for communication. Early in 2018, the UDP communication protocol was exploited with several amplification type DDoS attacks taking place on Memcached instances. Unlike TCP, UDP does not establish a handshake which reduces its overhead but leaves it vulnerable to attack. Popular uses of UDP protocols are VOIP and low-latency online games.  

It is important to disable UDP from your Memcached service in order to secure your [DEVICE] from amplification attacks. You can do this by blocking port 11211 which UDP uses. Alternatively, you can implement a firewall, but it is preferable to remove the need for UDP altogether. Reconfiguring Memcached to exclusively listen on the localhost is the safest option, although may not be realistic.

**Comment by fshmcallister** [comment]
Again, this might be a good one to do an article on!

**Comment by AlEsmail** [comment]
Change title: Memcached instance may *be* publicly accessible

**Comment by fshmcallister** [comment]
whoops, thank you

**Comment by vpetersson** [comment]
> Unlike TCP, UDP does not establish a handshake which reduces its overhead but leaves it vulnerable to attack. Popular uses of UDP protocols are VOIP and low-latency online games.

This isn't really relevant here. Also, it isn't true that UDP is vulnerable to this. The attack used was due to a number of factors, but the tl;dr for context was that you could send X bytes and have the server send X*N bytes to a spoofed address. More context [here](https://blog.cloudflare.com/memcrashed-major-amplification-attacks-from-port-11211/)

Anyway, we should probably recommend that they switch to TCP and disable UDP (unless they need UDP for some reason) as the remedy.

**Comment by fshmcallister** [comment]
**FAQ**
Memcached is a popular caching service used to help speed up websites. By default, Memcached enables TCP and UDP protocols for communication. Early in 2018, the UDP communication protocol was exploited with several amplification type DDoS attacks taking place on Memcached instances. Unlike TCP, UDP does not need to establish a handshake to send data which reduces its overhead but makes it easier to 'spoof.'

We recommend disabling UDP connections to your Memcached service in favour of TCP connections to secure your [DEVICE] from amplification attacks. You can do this by blocking port 11211 which UDP uses in the `/etc/sysconfig/memcached` file. Reconfiguring Memcached to exclusively listen on the localhost is the safest option, although may not be realistic if you wish to allow connections outside your local network. To secure connections from other [DEVICES], configure the Memcached instance to listen to a private IP. Note, these addresses vary by network.

**Comment by vpetersson** [comment]
Let me rewrite this - 

Memcached is a popular caching server. By default, Memcached enables TCP and UDP protocols for communication. Early in 2018, an exploit in Memcached's UDP communication was discovered. Using this exploit, an attacker was able to use a publicly exposed Memcached server to relay a payload and attack a 3rd party server/service, leading to large scale attacks across the internet.

We recommend disabling UDP connections to your Memcached service in favor of TCP connections to secure your server(s) from these attacks. You can do this by disabling UDP connections in Mamcache'd config file (often `/etc/sysconfig/memcached`). Re-configuring Memcached to exclusively listen on the localhost is the safest option, although may not be realistic if you wish to allow connections outside your local network. Depending on your network typology, you may be able to configure Memcached to listen on a private IP that is only accessible from the local network.

**Comment by fshmcallister** [comment]
Your Memcached instance may be accessible remotely. This is likely due to use of a default configuration. Disable UDP connections and traffic by blocking port 11211. Alternatively, reconfigure Memcached to only listen on localhost.

^ does that make sense in its current form?

**Comment by vpetersson** [comment]
Feel free to re-format it. I just wanted to reformat parts of it (e.g calling it a server and not service etc).

**Comment by fshmcallister** [comment]
Happy with that. Should we keep the code block separate by the way? Separated in the main text I wrote above but we can merge it into the FAQ if you feel it fits better there

**Comment by vpetersson** [comment]
@fshmcallister 

> Should we keep the code block separate by the way?

let's keep the code block in the "FAQ" section as you did.

> Separated in the main text I wrote above but we can merge it into the FAQ if you feel it fits better there

See above. But let's either update the top block or post a new complete  post with all the changes.

**Comment by fshmcallister** [comment]
**TITLE**
A Memcached instance may be publicly accessible 
> reconfigure Memcached 


**TL;DR**
Your Memcached instance may be accessible remotely. This is likely due to use of a default configuration. Disable UDP connections and traffic by blocking port 11211. Alternatively, reconfigure Memcached to only listen on localhost.

**FAQ**
Memcached is a popular caching server. By default, Memcached enables TCP and UDP protocols for communication. Early in 2018, an exploit in Memcached's UDP communication was discovered. Using this exploit, an attacker was able to use a publicly exposed Memcached server to relay a payload and attack a 3rd party server/service, leading to large scale attacks across the internet.

We recommend disabling UDP connections to your Memcached service in favor of TCP connections to secure your server(s) from these attacks. You can do this by disabling UDP connections in Mamcache'd config file (often `/etc/sysconfig/memcached`). Reconfiguring Memcached to exclusively listen on the localhost is the safest option, although may not be realistic if you wish to allow connections outside your local network. Depending on your network typology, you may be able to configure Memcached to listen on a private IP that is only accessible from the local network. 

To reconfigure, navigate to the following section in your Memcached config file:

```
PORT="11211"
USER="memcached"
MAXCONN="1024"
CACHESIZE="64"
OPTIONS="-l 127.0.0.1 -U 0"
```

Change the `OPTIONS` in Memcached config file to match the above. `-U 0` blocks all UDP and `-l 127.0.0.1` forces memcached to listen to localhost only.

Then restart:
```
$ sudo service memcached restart
```

This will now configure your memached instance to listen to the localhost and disables UDP traffic.
